File: frontend/project/lib/models/message.dart
class Message {
  final String id;
  final String requestId;
  final String senderId;
  final String receiverId;
  final String content;
  final bool isOffer;
  final double? offerAmount;
  final double? discountPercent;
  final double? discountedPrice;
  final DateTime createdAt;

  Message({
    required this.id,
    required this.requestId,
    required this.senderId,
    required this.receiverId,
    required this.content,
    this.isOffer = false,
    this.offerAmount,
    this.discountPercent,
    this.discountedPrice,
    required this.createdAt,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['_id'] ?? json['id'] ?? '',
      requestId: json['requestId'] ?? '',
      senderId: json['senderId'] ?? '',
      receiverId: json['receiverId'] ?? '',
      content: json['content'] ?? '',
      isOffer: json['isOffer'] ?? false,
      offerAmount: json['offerAmount']?.toDouble(),
      discountPercent: json['discountPercent']?.toDouble(),
      discountedPrice: json['discountedPrice']?.toDouble(),
      createdAt: json['createdAt'] != null 
          ? DateTime.parse(json['createdAt']) 
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'requestId': requestId,
      'senderId': senderId,
      'receiverId': receiverId,
      'content': content,
      'isOffer': isOffer,
      'offerAmount': offerAmount,
      'discountPercent': discountPercent,
      'discountedPrice': discountedPrice,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}

File: backend/src/models/Message.js
const mongoose = require('mongoose');

const messageSchema = new mongoose.Schema(
  {
    requestId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Request',
      required: true,
    },
    senderId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    receiverId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    content: {
      type: String,
      required: [true, 'Message content is required'],
      trim: true,
      maxlength: [1000, 'Message cannot exceed 1000 characters'],
    },
    isOffer: {
      type: Boolean,
      default: false,
    },
    offerAmount: {
      type: Number,
      default: null,
      min: 0,
    },
    discountPercent: {
      type: Number,
      default: 30,
      min: 0,
      max: 100,
    },
    discountedPrice: {
      type: Number,
      default: null,
      min: 0,
    },
    isRead: {
      type: Boolean,
      default: false,
    },
  },
  {
    timestamps: true,
  }
);

// Index for faster queries
messageSchema.index({ requestId: 1, createdAt: 1 });
messageSchema.index({ receiverId: 1, isRead: 1 });

module.exports = mongoose.model('Message', messageSchema);
